### DISCUSSION TITLE
uv_alloc_cb free list to ring buffer

### DISCUSSION THREAD

**Discussion by rynz** [libuv/libuv#4045] [Q&A]
Hi,

In the past, I have used a free list that implements something like:

```C
typedef struct {
  uv__queue items;
  unsigned int nitems;
} freelist_t

struct {
  uv_udp_send_t send_req
  char buf[1024];
  size_t len;
  uv__queue queue;
} freelist_item_t
```

Whereby I can grab a preallocated `freelist_item_t` in `uv_alloc_cb`, use its `buf`, send it with `uv_udp_send`, and then later return it in `uv_udp_send_cb` to the `freelist_t`.

I am reaching out to ask for your professional help:
Would using a ring buffer provide performance benefits as the initialised ring buffer would be contiguous memory?
As `uv_udp_send_cb` can be called out of order, would that create any issues with a ring buffer returning `freelist_item_t` out of order?

Any advice is greatly appreciated.

Cheers!

**Comment by vtjnash** [comment]
You have quite possibly just reinvented malloc/free and might be just as well off using those existing functions. Just curious if you have compared those?

**Reply by rynz** [reply]
Isn't there a performance penalty calling malloc every time we `uv_alloc_cb` rather than malloc a continuous block of memory at program initialisation and return pointers from it?

**Reply by vtjnash** [reply]
That malloc strategy is commonly known as an sbrk malloc or bump-pointer malloc, and it can perform excellently in certain cases, but it is also often fairly limited as it is a very primitive allocation strategy.

**Reply by rynz** [reply]
What allocation strategy would you recommend for use in `uv_alloc_cb`?
